test/common/lua/lua_test.cc

```diff
@@ -0,0 +1,153 @@
+
+  // Start a second coroutine but do hold a reference. Do a gc after finish which should not
+  // collect it. Then unref and collect and it should be gone.
+  CoroutinePtr cr2(state_->createCoroutine());
+  LuaRef<TestObject> ref2(TestObject::create(cr2->luaState()), true);
+  cr2->start(state_->getGlobalRef(1), 1, yield_callback_);
+  EXPECT_EQ(cr2->state(), Coroutine::State::Finished);
+  lua_gc(cr2->luaState(), LUA_GCCOLLECT, 0);
+  EXPECT_CALL(*ref2.get(), onDestroy());
+  ref2.reset();
+  lua_gc(cr2->luaState(), LUA_GCCOLLECT, 0);
+}
+
+// Basic yield/resume functionality.
+TEST_F(LuaTest, YieldAndResume) {
+  const std::string SCRIPT{R"EOF(
+    function callMe(object)
+      coroutine.yield()
+      object:testCall()
+    end
+  )EOF"};
+
+  InSequence s;
+  setup(SCRIPT);
+  EXPECT_NE(LUA_REFNIL, state_->getGlobalRef(state_->registerGlobal("callMe")));
+
+  CoroutinePtr cr(state_->createCoroutine());
+  LuaRef<TestObject> ref(TestObject::create(cr->luaState()), true);
+  EXPECT_CALL(on_yield_, ready());
+  cr->start(state_->getGlobalRef(0), 1, yield_callback_);
+  EXPECT_EQ(cr->state(), Coroutine::State::Yielded);
+
+  EXPECT_CALL(*ref.get(), doTestCall(_));
+  cr->resume(0, yield_callback_);
+  EXPECT_EQ(cr->state(), Coroutine::State::Finished);
+
+  lua_gc(cr->luaState(), LUA_GCCOLLECT, 0);
+  EXPECT_CALL(*ref.get(), onDestroy());
+  ref.reset();
+  lua_gc(cr->luaState(), LUA_GCCOLLECT, 0);
+}
+
+// Mark dead/live and ref counting across coroutines.
+TEST_F(LuaTest, MarkDead) {
+  const std::string SCRIPT{R"EOF(
+    function callMeFirst(object)
+      global_object = object
+      global_object:testCall()
+      coroutine.yield()
+      global_object:testCall()
+    end
+
+    function callMeSecond()
+      global_object:testCall()
+    end
+  )EOF"};
+
+  InSequence s;
+  setup(SCRIPT);
+  EXPECT_NE(LUA_REFNIL, state_->getGlobalRef(state_->registerGlobal("callMeFirst")));
+  EXPECT_NE(LUA_REFNIL, state_->getGlobalRef(state_->registerGlobal("callMeSecond")));
+
+  CoroutinePtr cr1(state_->createCoroutine());
+  LuaDeathRef<TestObject> ref(TestObject::create(cr1->luaState()), true);
+  EXPECT_CALL(*ref.get(), doTestCall(_));
+  EXPECT_CALL(on_yield_, ready());
+  cr1->start(state_->getGlobalRef(0), 1, yield_callback_);
+  EXPECT_EQ(cr1->state(), Coroutine::State::Yielded);
+
+  ref.markDead();
+  CoroutinePtr cr2(state_->createCoroutine());
+  EXPECT_THROW_WITH_MESSAGE(cr2->start(state_->getGlobalRef(1), 0, yield_callback_), LuaException,
+                            "[string \"...\"]:10: object used outside of proper scope");
+  EXPECT_EQ(cr2->state(), Coroutine::State::Finished);
+
+  ref.markLive();
+  EXPECT_CALL(*ref.get(), doTestCall(_));
+  cr1->resume(0, yield_callback_);
+  EXPECT_EQ(cr1->state(), Coroutine::State::Finished);
+
+  lua_gc(cr1->luaState(), LUA_GCCOLLECT, 0);
+  EXPECT_CALL(*ref.get(), onDestroy());
+  ref.reset();
+  lua_gc(cr1->luaState(), LUA_GCCOLLECT, 0);
+}
+
+} // namespace Lua
+} // namespace Envoy
```

Comment:
Would unit tests for `Lua[Death]Ref` make sense?

Reply:
These are covered already in these tests. I'm not sure how to do this without mocking the central Lua state. I can look into this.

Reply:
I think I'm going to skip this for now as it would require fully mocking the Lua state. I added some TODOs.